evm/transaction_context.py

```diff
@@ -0,0 +1,28 @@
+from evm.validation import (
+    validate_canonical_address,
+    validate_uint256,
+)
+
+
+class BaseTransactionContext(object):
+    """
+    This immutable object houses information that remains constant for the entire context of the VM
+    execution.
+    """
+    _gas_price = None
+    _origin = None
```

Comment:
I don't think we need the `@property` versions of these.  Should be good just using them directly as properties.

Reply:
I copied the behavior of https://github.com/hwwhww/py-evm/blob/66a858f6ed1b4a02e35c311ed3255440499c8f9a/evm/vm/execution_context.py. Should this be different for some reason?

Reply:
The reason for why we protect fields of `ExecutionContext` is that we want to make sure the fields **only be assigned in `__init__`** and won't be changed in other places.
But I'm not sure what's the use cases of `TransactionContext`.

Reply:
@pipermerriam originally said:

> This would be an object that we treat as immutable

So it should be the same, no?

Reply:
@hwwhww `@property` methods don't actually provide any protection from mutation *unless* you create a `setter` which throws.  I'd be good with doing that but it should probably be done with our own decorator that automatically creates the `setter` function.  Something like a new decorator:

```python
@immutable_property
def something(self):
    ...
```

Reply:
@pipermerriam 
So sorry for that I just realized I missed and haven't seen this comment(https://github.com/ethereum/py-evm/pull/273#discussion_r161866203) 15 days ago! Agree with you about using explicit `@read_only_property` or `@immutable_property` decorator.

Reply:
No big deal.  Easy enough for us to make that cleanup now (mind opening an issue for it?)

@lrettig I'm fine with leaving current `@property` methods in place and cleaning them up with a stand-alone PR with the `@immutable_property/read_only_property` decorator.

Reply:
Sounds good, thanks for the clarification